### PR TITLE
Small workbench-lib improvements - akka idle timeout

### PR DESCRIPTION
**Pull request by aweng98**
**PR checklist**
- [ ] For each library you've modified here, decide whether it requires a major, minor, or no version bump. (Click [here](/broadinstitute/workbench-libs/blob/develop/CONTRIBUTING.md) for guidance)

If you're doing a **major** or **minor** version bump to any libraries:
- [ ] Bump the version in `project/Settings.scala` `createVersion()`
- [ ] Update `CHANGELOG.md` for those libraries
- [ ] I promise I used `@deprecated` instead of deleting code where possible

In all cases:
- [ ] Replace the appropriate version hashes in `README.md` and the `CHANGELOG.md` for any libs you changed with `TRAVIS-REPLACE-ME` to auto-update the version string
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green (CI _and_ coverage tests)
- [ ] Squash commits and **merge to develop**
- [ ] Delete branch after merge

### CODE REVIEW

File: serviceTest/src/test/resources/application.conf [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  # Loggers to register at boot time (akka.event.Logging$DefaultLogger logs
+  # to STDOUT)
+  loggers = ["akka.event.Logging$DefaultLogger"]
+  
+  # Options: OFF, ERROR, WARNING, INFO, DEBUG
+  loglevel = "INFO"
+
+  # Log the complete configuration at INFO level when the actor system is started.
+  # Useful to know configuration.
+  log-config-on-start = on
```
Comment (aweng98) [inline serviceTest/src/test/resources/application.conf]: used this setting to see all akka test config during PR testing. very useful. I will turn it off before final merge.

File: serviceTest/src/test/resources/application.conf [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        parallelism-factor = 2.0
+        # Max number of threads to cap factor-based parallelism number to
+        parallelism-max = 8
+      }
+      # Throughput defines the maximum number of messages to be
+      # processed per actor before the thread jumps to the next actor.
+      throughput = 1
+    }
+
+    blocking-dispatcher {
```
Comment (aweng98) [inline serviceTest/src/test/resources/application.conf]: akka Trick and Tips: How to handle blocking api calls.
https://doc.akka.io/docs/akka-http/current/handling-blocking-operations-in-akka-http-routes.html

File: serviceTest/src/test/resources/application.conf [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        fixed-pool-size = 8
+      }
+      throughput = 1
+    }
+  }
+  
+  http {
+    client {
+      # The time after which an idle connection will be automatically closed.
+      idle-timeout = 600 s
```
Comment (aweng98) [inline serviceTest/src/test/resources/application.conf]: 5 minutes. match `RestClient.sendRequest` blocking call `Await.result(responseFuture, 5.minutes)`.
https://github.com/broadinstitute/workbench-libs/blob/4d3298807e8fded585277eafd4995e513091105b/serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/RestClient.scala#L63

**Comment by aweng98** [comment]
small improvements for workbench-lib Service that help with test execution and make issue troubleshooting easier.
1:  ~new workbench-lib logs~
2: increase akka idle timeout

Without change, an error `TCP idle-timeout encountered` captured in test log.
```
14:46:12.652 [default-akka.actor.default-dispatcher-7]          Rawls$:applyOrElse             INFO  retry-able operation failed: 6 retries remaining, retrying in 1667455552 nanoseconds
akka.stream.scaladsl.TcpIdleTimeoutException: TCP idle-timeout encountered on connection to [rawls-fiab.dsde-qa.broadinstitute.org:24443], no bytes passed in the last 1 minute
	at akka.stream.impl.io.TcpIdleTimeout$$anonfun$1.applyOrElse(TcpStages.scala:374)
	at akka.stream.impl.io.TcpIdleTimeout$$anonfun$1.applyOrElse(TcpStages.scala:372)
```	
	
Can't really tell what the application error is. With a bigger akka idle timeout. now we know real error is `"message" : "The project to be billed is associated with an absent billing account."`
```
[INFO ] 2018-09-05 15:28:37.169 [workbenchServiceTest-akka.actor.blocking-dispatcher-55] Rawls$:applyOrElse             - retry-able operation failed: 6 retries remaining, retrying in 1921629952 nanoseconds
org.broadinstitute.dsde.workbench.service.RestException: {"source":"rawls","causes":[],"exceptionClass":"com.google.api.client.googleapis.json.GoogleJsonResponseException","stackTrace":[{"className":"com.google.api.client.googleapis.json.GoogleJsonResponseException","methodName":"from","fileName":"GoogleJsonResponseException.java","lineNumber":146},{"className":"com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest","methodName":"newExceptionOnError","fileName":"AbstractGoogleJsonClientRequest.java","lineNumber":113},{"className":"com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest","methodName":"newExceptionOnError","fileName":"AbstractGoogleJsonClientRequest.java","lineNumber":40},{"className":"com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1","methodName":"interceptResponse","fileName":"AbstractGoogleClientRequest.java","lineNumber":321},{"className":"com.google.api.client.http.HttpRequest","methodName":"execute","fileName":"HttpRequest.java","lineNumber":1065},{"className":"com.google.api.client.googleapis.services.AbstractGoogleClientRequest","methodName":"executeUnparsed","fileName":"AbstractGoogleClientRequest.java","lineNumber":419},{"className":"com.google.api.client.googleapis.services.AbstractGoogleClientRequest","methodName":"executeUnparsed","fileName":"AbstractGoogleClientRequest.java","lineNumber":352},{"className":"org.broadinstitute.dsde.rawls.google.GoogleUtilities","methodName":"$anonfun$executeGoogleCall$1","fileName":"GoogleUtilities.scala","lineNumber":63},{"className":"scala.util.Try$","methodName":"apply","fileName":"Try.scala","lineNumber":209},{"className":"org.broadinstitute.dsde.rawls.google.GoogleUtilities","methodName":"executeGoogleCall","fileName":"GoogleUtilities.scala","lineNumber":63},{"className":"org.broadinstitute.dsde.rawls.google.GoogleUtilities","methodName":"executeGoogleCall$","fileName":"GoogleUtilities.scala","lineNumber":60},{"className":"org.broadinstitute.dsde.rawls.dataaccess.HttpGoogleServicesDAO","methodName":"executeGoogleCall","fileName":"HttpGoogleServicesDAO.scala","lineNumber":59},{"className":"org.broadinstitute.dsde.rawls.google.GoogleUtilities","methodName":"executeGoogleRequest","fileName":"GoogleUtilities.scala","lineNumber":44},{"className":"org.broadinstitute.dsde.rawls.google.GoogleUtilities","methodName":"executeGoogleRequest$","fileName":"GoogleUtilities.scala","lineNumber":43},{"className":"org.broadinstitute.dsde.rawls.dataaccess.HttpGoogleServicesDAO","methodName":"executeGoogleRequest","fileName":"HttpGoogleServicesDAO.scala","lineNumber":59},{"className":"org.broadinstitute.dsde.rawls.dataaccess.HttpGoogleServicesDAO","methodName":"$anonfun$grantReadAccess$3","fileName":"HttpGoogleServicesDAO.scala","lineNumber":327},{"className":"scala.collection.immutable.List","methodName":"flatMap","fileName":"List.scala","lineNumber":335},{"className":"org.broadinstitute.dsde.rawls.dataaccess.HttpGoogleServicesDAO","methodName":"$anonfun$grantReadAccess$2","fileName":"HttpGoogleServicesDAO.scala","lineNumber":326},{"className":"scala.collection.TraversableLike","methodName":"$anonfun$flatMap$1","fileName":"TraversableLike.scala","lineNumber":241},{"className":"scala.collection.immutable.Set$Set2","methodName":"foreach","fileName":"Set.scala","lineNumber":130},{"className":"scala.collection.TraversableLike","methodName":"flatMap","fileName":"TraversableLike.scala","lineNumber":241},{"className":"scala.collection.TraversableLike","methodName":"flatMap$","fileName":"TraversableLike.scala","lineNumber":238},{"className":"scala.collection.AbstractTraversable","methodName":"flatMap","fileName":"Traversable.scala","lineNumber":104},{"className":"org.broadinstitute.dsde.rawls.dataaccess.HttpGoogleServicesDAO","methodName":"insertNewAcls$1","fileName":"HttpGoogleServicesDAO.scala","lineNumber":317},{"className":"org.broadinstitute.dsde.rawls.dataaccess.HttpGoogleServicesDAO","methodName":"$anonfun$grantReadAccess$5","fileName":"HttpGoogleServicesDAO.scala","lineNumber":331},{"className":"scala.concurrent.BlockContext$","methodName":"withBlockContext","fileName":"BlockContext.scala","lineNumber":81},{"className":"scala.concurrent.impl.ExecutionContextImpl$DefaultThreadFactory$$anon$2$$anon$5","methodName":"block","fileName":"ExecutionContextImpl.scala","lineNumber":76},{"className":"java.util.concurrent.ForkJoinPool","methodName":"managedBlock","fileName":"ForkJoinPool.java","lineNumber":3313},{"className":"scala.concurrent.impl.ExecutionContextImpl$DefaultThreadFactory$$anon$2","methodName":"blockOn","fileName":"ExecutionContextImpl.scala","lineNumber":71},{"className":"scala.concurrent.package$","methodName":"blocking","fileName":"package.scala","lineNumber":142},{"className":"org.broadinstitute.dsde.rawls.google.GoogleUtilities","methodName":"$anonfun$retryWithRecoverWhen500orGoogleError$3","fileName":"GoogleUtilities.scala","lineNumber":40},{"className":"scala.concurrent.Future$","methodName":"$anonfun$apply$1","fileName":"Future.scala","lineNumber":653},{"className":"scala.util.Success","methodName":"$anonfun$map$1","fileName":"Try.scala","lineNumber":251},{"className":"scala.util.Success","methodName":"map","fileName":"Try.scala","lineNumber":209},{"className":"scala.concurrent.Future","methodName":"$anonfun$map$1","fileName":"Future.scala","lineNumber":287},{"className":"scala.concurrent.impl.Promise","methodName":"liftedTree1$1","fileName":"Promise.scala","lineNumber":29},{"className":"scala.concurrent.impl.Promise","methodName":"$anonfun$transform$1","fileName":"Promise.scala","lineNumber":29},{"className":"scala.concurrent.impl.CallbackRunnable","methodName":"run","fileName":"Promise.scala","lineNumber":60},{"className":"scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask","methodName":"exec","fileName":"ExecutionContextImpl.scala","lineNumber":140},{"className":"java.util.concurrent.ForkJoinTask","methodName":"doExec","fileName":"ForkJoinTask.java","lineNumber":289},{"className":"java.util.concurrent.ForkJoinPool$WorkQueue","methodName":"pollAndExecAll","fileName":"ForkJoinPool.java","lineNumber":1021},{"className":"java.util.concurrent.ForkJoinPool$WorkQueue","methodName":"execLocalTasks","fileName":"ForkJoinPool.java","lineNumber":1046},{"className":"java.util.concurrent.ForkJoinPool$WorkQueue","methodName":"runTask","fileName":"ForkJoinPool.java","lineNumber":1058},{"className":"java.util.concurrent.ForkJoinPool","methodName":"runWorker","fileName":"ForkJoinPool.java","lineNumber":1692},{"className":"java.util.concurrent.ForkJoinWorkerThread","methodName":"run","fileName":"ForkJoinWorkerThread.java","lineNumber":157}],"message":"403 Forbidden\n{\n  \"code\" : 403,\n  \"errors\" : [ {\n    \"domain\" : \"global\",\n    \"location\" : \"Authorization\",\n    \"locationType\" : \"header\",\n    \"message\" : \"The project to be billed is associated with an absent billing account.\",\n    \"reason\" : \"accountDisabled\"\n  } ],\n  \"message\" : \"The project to be billed is associated with an absent billing account.\"\n}"}
	at org.broadinstitute.dsde.workbench.service.RestClient.throwRestException(RestClient.scala:100)
	at org.broadinstitute.dsde.workbench.service.RestClient.$anonfun$sendRequest$2(RestClient.scala:75)
```

**Comment by aweng98** [comment]
log not created. I don't know why at this time.

**Comment by aweng98** [comment]
speculative  fix for https://broadinstitute.atlassian.net/browse/GAWB-3750. Because Orchestration api request will wait with sufficient time until Rawls retries are complete. The test won't terminate and clean up (release claimed billing project, delete groups, or/and delete workspaces) prematurely while Rawls is still trying update groups for the workspace create.

File: serviceTest/src/test/resources/application.conf (hunk 3 above)
Comment (ahaessly) [reply serviceTest/src/test/resources/application.conf]: did you want to wait 5 minutes (300s) or 10 mins (600s)?

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/RestClient.scala [hunk 4]
```diff
@@ -49,7 +51,7 @@ trait RestClient extends Retry with LazyLogging {
 
   private def sendRequest(httpRequest: HttpRequest): HttpResponse = {
     val responseFuture = retryExponentially() {
-      () => Http().singleRequest(httpRequest).map { response =>
+      () => Http().singleRequest(request = httpRequest).map { response =>
```
Comment (ahaessly) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/RestClient.scala:54]: is the request =  a significant change? just wondering why

**Review by ahaessly — APPROVED** [review]
see comments

**Comment by ahaessly** [comment]
i'm guessing the logging didn't work because the application that imports the workbench libs and actually runs probably has it's own logging configuration that has precedence

File: serviceTest/src/test/resources/application.conf (hunk 3 above)
Comment (aweng98) [reply serviceTest/src/test/resources/application.conf]: Uh, oops. Good Catch!

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/RestClient.scala (hunk 4 above)
Comment (aweng98) [reply serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/RestClient.scala:54]: No change. Using named parameter in method calling.

File: serviceTest/src/test/resources/application.conf (hunk 2 above)
Comment (kyuksel) [reply serviceTest/src/test/resources/application.conf]: Have we run into the starvation issue mentioned on that link when not using a `blocking-dispatcher`?

**Review by kyuksel — COMMENTED** [review]
I'm fine with increasing the idle timeout. I'd like to understand what specific issue led to the dedicated dispatcher.

**Comment by aweng98** [comment]
@kyuksel In example that I gave earlier, Rawls is retrying Google api that take longer than a minute to end. While retrying, Rawls does not give out any api response. Orchestration api idle time out default is set at 60 seconds. Resulting Orch. api time out before hearing back from Rawls. Rawls exponential retry intervals are: 1000 milliseconds, 2000 milliseconds, 4000 milliseconds, 8000 milliseconds, 16000 milliseconds, 32000 milliseconds

**Comment by kyuksel** [comment]
@aweng98 Why do we need the change below, though? How is it related to the timeout issue?

```
actor {
     # Dedicated dispatcher for blocking operations
     # https://doc.akka.io/docs/akka/2.4.11/scala/http/handling-blocking-operations-in-akka-http-routes.html
     blocking-dispatcher {
       type = Dispatcher
       executor = "thread-pool-executor"
       thread-pool-executor {
         fixed-pool-size = 16
       }
       throughput = 100
     }
```

**Comment by aweng98** [comment]
Something to improve akka http requests to prevent having too many idling and locking threads while waiting for Rawls to respond. Because `Http()` call is done synchronously waiting `Await.result(responseFuture, 5.minutes)`. The tests in all services are running concurrently.

**Comment by kyuksel** [comment]
Ah, I see 👍

**Review by kyuksel — APPROVED** [review]

**Pull request merged**